source/extensions/filters/udp/udp_proxy/session_filters/http_capsule/http_capsule.cc

```diff
@@ -0,0 +1,86 @@
+#include "source/extensions/filters/udp/udp_proxy/session_filters/http_capsule/http_capsule.h"
+
+#include "source/common/common/hex.h"
+
+#include "absl/strings/escaping.h"
+#include "quiche/common/masque/connect_udp_datagram_payload.h"
+#include "quiche/common/simple_buffer_allocator.h"
+#include "quiche/quic/core/http/quic_spdy_stream.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace UdpProxy {
+namespace SessionFilters {
+namespace HttpCapsule {
+
+ReadFilterStatus HttpCapsuleFilter::onData(Network::UdpRecvData& data) {
+  std::string buffer = data.buffer_->toString();
+  quiche::ConnectUdpDatagramUdpPacketPayload payload(buffer);
+  quiche::QuicheBuffer serialized_capsule =
+      SerializeCapsule(quiche::Capsule::Datagram(payload.Serialize()), &capsule_buffer_allocator_);
+
+  data.buffer_->drain(data.buffer_->length());
+  data.buffer_->add(serialized_capsule.AsStringView());
+  return ReadFilterStatus::Continue;
+}
+
+WriteFilterStatus HttpCapsuleFilter::onWrite(Network::UdpRecvData& data) {
+  // TODO(ohadvano): add filter callbacks to get addresses instead of saving them.
+  local_address_ = data.addresses_.local_;
+  peer_address_ = data.addresses_.peer_;
+
+  for (const Buffer::RawSlice& slice : data.buffer_->getRawSlices()) {
+    absl::string_view mem_slice(static_cast<const char*>(slice.mem_), slice.len_);
+    if (!capsule_parser_.IngestCapsuleFragment(mem_slice)) {
+      ENVOY_LOG(error, "Capsule ingestion error occured: slice = {}", mem_slice);
```

Comment:
I wouldn't log `mem_slice` here as it's not guaranteed to be encoded as UTF-8. I'd suggest only logging `slice.len_` instead